### PR TITLE
Add checks on SHR's src1 during UnsignedCmpPeep in FGPeeps

### PR DESCRIPTION
**Pull request by meg-gupta**

### CODE REVIEW

**Comment by meg-gupta** [comment]
@rajatd Can you please review ?

File: lib/Backend/FlowGraph.cpp [hunk 1]
```diff
@@ -3014,7 +3014,15 @@ bool FlowGraph::IsUnsignedOpnd(IR::Opnd *src, IR::Opnd **pShrSrc1)
         return false;
     }
 
-    *pShrSrc1 = shrUInstr->GetSrc1();
+    IR::Opnd *shrSrc = shrUInstr->GetSrc1();
+
+    Assert (shrSrc->IsRegOpnd());
+    if (!shrSrc->AsRegOpnd()->m_sym->m_isIntConst)
```
Comment (ThomsonTan) [inline lib/Backend/FlowGraph.cpp]: >if (!shrSrc->AsRegOpnd()->m_sym->m_isIntConst) [](start = 4, length = 46)

Could we made a test case for this?

Comment (meg-gupta) [inline lib/Backend/FlowGraph.cpp]: Added testcase.

**Comment by MikeHolman** [comment]
Why are we doing an |0? Semantics of js are that |0 converts to signed. The point of the  >>> 0 in source is to convert to unsigned.

File: lib/Backend/FlowGraph.cpp [hunk 2]
```diff
@@ -3014,7 +3014,11 @@ bool FlowGraph::IsUnsignedOpnd(IR::Opnd *src, IR::Opnd **pShrSrc1)
         return false;
     }
 
-    *pShrSrc1 = shrUInstr->GetSrc1();
+    IR::Opnd *shrSrc = shrUInstr->GetSrc1();
+
+    Assert (shrSrc->IsRegOpnd());
```
Comment (ThomsonTan) [inline lib/Backend/FlowGraph.cpp]: >Assert (shrSrc->IsRegOpnd()); [](start = 4, length = 29)

Why assert here, got some violation?

File: lib/Backend/FlowGraph.cpp [hunk 3]
```diff
@@ -3034,6 +3038,7 @@ bool FlowGraph::UnsignedCmpPeep(IR::Instr *cmpInstr)
     //
     //  t1 = ShrU_A x, 0
     //  t2 = 10;
+    //  t3 = Or_A x, 0
     //  BrUnGt x, t2, L
```
Comment (ThomsonTan) [inline lib/Backend/FlowGraph.cpp]: >x [](start = 15, length = 1)

Sounds like this should be `BrUnGt t3, t2, L`.

**Comment by ThomsonTan** [comment]
I guess `Or_A` is not exactly the same as `|` in JavaScript, and it should be fine to calculate temporary result because its result is not affected by signed/unsigned. Or we probably could use `MOV_TRUNK`?

---
In reply to: [283729990](https://github.com/Microsoft/ChakraCore/pull/2601#issuecomment-283729990) [](ancestors = 283729990)

**Comment by meg-gupta** [comment]
@MikeHolman We do this optimization only when one of the compare branch's sources is a non negative int constant. | 0 was done to convert to 32-bits. Since we are converting signed GT/GE/LT/LE to unsigned compare branches, we produce correct results. For signed EQ/NEQ compare branches, we produce correct results because one of the operands is always non negative int constant.

File: lib/Backend/FlowGraph.cpp (hunk 3 above)
Comment (meg-gupta) [inline lib/Backend/FlowGraph.cpp]: Yes, will fix this.

**Comment by meg-gupta** [comment]
@dotnet-bot test Ubuntu static_ubuntu_linux_release

**Comment by MikeHolman** [comment]
I looked into this more, and I don't really like this peep... It seems like a hack to tell globopt that `a >>> 0` will make dst into a uint (which we don't support well). But, this fixes the bug, so LGTM.

**Pull request merged**